cmd/zdb/zdb.c

```diff
@@ -6346,6 +6347,34 @@ zdb_parse_block_sizes(char *sizes, uint64_t *lsize, uint64_t *psize)
 	return (*lsize >= *psize && *psize > 0);
 }
 
+static boolean_t
+zdb_blkptr_verify(const blkptr_t *bp)
+{
+	if (!DMU_OT_IS_VALID(BP_GET_TYPE(bp))) {
+		return (B_FALSE);
+	}
+	if (BP_GET_CHECKSUM(bp) >= ZIO_CHECKSUM_FUNCTIONS ||
+	    BP_GET_CHECKSUM(bp) <= ZIO_CHECKSUM_ON) {
+		return (B_FALSE);
+	}
+	if (BP_GET_COMPRESS(bp) >= ZIO_COMPRESS_FUNCTIONS ||
+	    BP_GET_COMPRESS(bp) <= ZIO_COMPRESS_ON) {
+		return (B_FALSE);
+	}
+	if (BP_GET_LSIZE(bp) > SPA_MAXBLOCKSIZE) {
+		return (B_FALSE);
+	}
+	if (BP_GET_PSIZE(bp) > SPA_MAXBLOCKSIZE) {
+		return (B_FALSE);
+	}
+	if (BP_IS_EMBEDDED(bp)) {
+		if (BPE_GET_ETYPE(bp) >= NUM_BP_EMBEDDED_TYPES) {
+			return (B_FALSE);
+		}
+	}
```

Comment:
It would be nice to include the DVA checks from `zfs_blkptr_verify` here.  Or possibly better yet, do a little refactoring so there's a single shared function to validate bps.  This could be left as a follow up improvement.